### ISSUE TITLE
Coerce dependencies using Array() and document it as being required for adapters

### ISSUE THREAD

**Issue by ericfranz** [OSC/ood_job#3]
https://github.com/OSC/ood_job/blob/ad8ded0c070157fb7121b557382d262c52c3480a/lib/ood_job/adapters/torque.rb#L45-L48

```ruby
        after      = [after].flatten.map(&:to_s)
        afterok    = [afterok].flatten.map(&:to_s)
        afternotok = [afternotok].flatten.map(&:to_s)
        afterany   = [afterany].flatten.map(&:to_s)
```

to

```ruby
        after      = Array(after).map(&:to_s)
        etc.
```

And document this as required for all adapters.

**Comment by nickjer** [comment]
This doesn't work for hashes:

```ruby
# Our argument is a hash...
arg = {a: 1}

# Original...
[ arg ].flatten #=> [{:a => 1}]

# Using Array...
Array( arg ) #=> [[:a, 1]]
```

**Comment by ericfranz** [comment]
In what case is afterok, after, afternotok, etc. a hash? shouldn't it be 1 id or an array of ids?

**Comment by ericfranz** [comment]
Documentation specifies this can only be a String or an array of Strings, not a hash: https://github.com/OSC/ood_job/blob/ad8ded0c070157fb7121b557382d262c52c3480a/lib/ood_job/adapter.rb#L32-L40

**Comment by nickjer** [comment]
It occurs in the initialization of `OodJob::Script` for `nodes`. I don't see the problem with:

```ruby
[arg].flatten
```

as it reproduces `Array()` and doesn't have the hidden gotcha for hashes. I'd also rather keep the array transformation consistent throughout the code so as to not confuse people why it is one way here but another way elsewhere.

**Comment by ericfranz** [comment]
Actually u think it's superfluous and actually incorrect here to do

    after      = [after].flatten.map(&:to_s)

Instead of just

    after      = [after].flatten

The method signature says the adapter needs to support only String or an array of Strings, which are ids returned by the submit method. Not another object that by calling to_s on it would return corresponding job id.

Also this would actually be inconsistent because the other adapter methods  claim they only accept a string for job id as an argument, and we do not coerce the passed in object to a string.

While it would be clever to be able to define #to_s on a Job object to return an id and then be able to do something like adapter.status(id: jobobj) I think then it's still better to do adapter.status(id: jobobj.id) and in the same way instead of adapter.submit(script: postscript, after: jobobj) to do require adapter.submit(script: postscript, after: jobobj.id)

But in most cases it seems that we would never have this scenario. Instead, we submit a job, get a return value, and pass that argument right back to submit as a dependency. And that will always be a string or array of strings.

**Comment by ericfranz** [comment]
And in that case I still think you should coerce these arguments to an Array using Array() because it's clearer what you are doing (and it's a common practice) and that's the code that would need to be implemented by someone else when building a new adapter (if that ever happens) whereas the hash coercion is a special case that occurs in Script class that no one is likely to touch.

**Comment by ericfranz** [comment]
In your favor: http://guides.rubyonrails.org/active_support_core_extensions.html#wrapping

**Comment by ericfranz** [comment]
http://api.rubyonrails.org/classes/Array.html#method-c-wrap

```ruby
def self.wrap(object)
  if object.nil?
    []
  elsif object.respond_to?(:to_ary)
    object.to_ary || [object]
  else
    [object]
  end
end
```

**Comment by ericfranz** [comment]
And to be even more pedantic:

```ruby
a = [[1,2],[3,4],[5,6]]
[a].flatten #=> [1, 2, 3, 4, 5, 6] oops!
[a].flatten(1) #=> [[1, 2], [3, 4], [5, 6]] better, if thats what you're going for

b = [1,2]
[b].flatten(1) #=> [1,2] but you get this no matter what method of coercion
               #  you choose, for obvious reasons
```

**Comment by nickjer** [comment]
I can just change the Ruby docs to say it accepts `#to_s` instead of `String`. I only did it `String` before because I wanted to make it easy on the developers, but since we are doing the array conversion anyways, then that argument is moot.

I did see the `flatten(1)` in StackOverflow, but didn't implement it because there was no case of nested arrays in this code, but there are cases of hashes. I am fine with padding all `flatten` to `flatten(1)`. Although I am not sure about adding ActiveSupport as a dependency.
